src/main/java/com/google/step/snippet/external/StackOverflowClient.java

```diff
@@ -0,0 +1,149 @@
+package com.google.step.snippet.external;
+
+import com.google.step.snippet.data.Card;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.*;
+import java.util.regex.Pattern;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.json.JSONObject;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+public final class StackOverflowClient {
+  private static final String SEARCH_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s?"
+          + "order=desc&sort=activity&site=stackoverflow";
+  private static final String QUESTION_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/questions/%s/answers?"
+          + "order=desc&sort=votes&site=stackoverflow";
+  // This URL specifies a custom StackExchange API filter that generates answer body.
+  private static final String ANSWER_URL_TEMPLATE =
+      "https://api.stackexchange.com/2.2/answers/%s?order"
+          + "=desc&sort=activity&site=stackoverflow&filter=!9_bDE(fI5";
+  // The URL is in the pattern of stackoverlow.com/questions/question_id/title.
+  // The ID_INDEX help retrieve the question_id from parsed URL.
+  private static final int ID_INDEX = 2;
+  private static final String ITEM_PARAMETER = "items";
+  private static final String TITLE_PARAMETER = "title";
+  private static final String BODY_PARAMETER = "body";
+  private static final String CODE_PARAMETER = "code";
+  private static final String ANSWER_ID_PARAMETER = "answer_id";
+  // Set 200 to be the maximum length of description for MVP.
+  private static final int DESCRIPTION_LENGTH_PARAMETER = 200;
+
+  /* This method build the desired card after getting each field. */
+  /* It will return a null card if no valid question is found */
+  public Card search(String url) {
+    String questionId;
+    try {
+      questionId = getQuestionId(url);
+    } catch (URISyntaxException e) {
+      // Return null card if no valid card available.
+      return null;
+    }
+    String answerId = getAnswerId(questionId);
+    if (answerId == null) {
+      return null;
+    }
+    String title = getTitle(questionId);
+    String answerBody = getAnswerBody(answerId);
+    String description = getDescription(answerBody);
+    String code = getCode(answerBody);
+    return new Card(title, code, url, description);
```

Comment:
In Kayla's W3Schools parser, Cards were only returned if all fields are non-empty. The same check isn't done here - is that intentional?

Reply:
yes, since stackoverflow sometimes has answer that is without code snippet/description, so the same check will not happen here